File: AirSim/project/airsim_2dnav/scripts/cpp/efficiency.cpp
#include "ros/ros.h"
#include "geometry_msgs/Vector3.h"
#include "std_msgs/Float64.h"
#include "std_msgs/Float64MultiArray.h"
#include "std_msgs/Int8.h"
#include "nav_msgs/Odometry.h"
#include <vector>
using std::vector;

class Efficiency
{
private:
  // Constants
  const double mass = 2176.46; // mass of the vehicle [kg]
  const double g = 9.81; // gravity constant [m/s^2]
  const double friction = 0.01; // rolling friction
  const double rho = 1.2; // for air at NTP [kg/m^3]
  const double drag_coeff = 0.357; // drag coefficient
  const double area = 3.389; // front area of vehicle [m^2]
  const double f_r = mass*g*friction;
  const double drag = 0.5 * rho * drag_coeff * area;

  // Goal
  vector<vector<double>> goals = {{44.50, -65.75}, {0, -256}, {-84.5,-198.22}, {-135, 1.50}};
  const int last_goal_int = 3; // Change last goal here
  double last_goal[2] = {goals[last_goal_int][0], goals[last_goal_int][1]};

  // Graph
  vector<double> v_array, a_array, e_array, d_array, ma_array, t_array, goaltime;
  int passed;

  // Variables
  double distance, energy, duration, x0, y0, z0, v0, v_x0, v_y0, v_z0, t0;
  bool start, end, showgraph;
  ros::Publisher pub_v, pub_a, pub_e, pub_d, pub_ma, pub_t, pub_gt;
  std_msgs::Float64MultiArray v_dat, a_dat,e_dat, d_dat, ma_dat, t_dat, gt_dat;
  ros::Subscriber sub_odom, sub_score;
public:
  Efficiency(ros::NodeHandle *n) {
    v0 = v_x0 = v_y0 = v_z0 = x0 = y0 = z0 = t0 = distance = duration = energy = 0;
    start = end = false;
    passed = 0;
    showgraph = true;

    pub_v = n->advertise<std_msgs::Float64MultiArray>("/velocity", 1);
    pub_a = n->advertise<std_msgs::Float64MultiArray>("/acceleration", 1);
    pub_e = n->advertise<std_msgs::Float64MultiArray>("/energy", 1);
    pub_d = n->advertise<std_msgs::Float64MultiArray>("/dragF", 1);
    pub_ma = n->advertise<std_msgs::Float64MultiArray>("/inertialF", 1);
    pub_gt = n->advertise<std_msgs::Float64MultiArray>("/goalTime", 1);
    pub_t = n->advertise<std_msgs::Float64MultiArray>("/time", 1);
    sub_score = n->subscribe("/current_score", 1, &Efficiency::getScore, this);
    sub_odom = n->subscribe("/odom", 1000, &Efficiency::odom, this);
  }
  void odom(const nav_msgs::Odometry::ConstPtr& msg);
  void getScore(const std_msgs::Int8::ConstPtr& num);
};

void Efficiency::odom(const nav_msgs::Odometry::ConstPtr &msg) {
  double car_x = msg->pose.pose.position.x;
  double car_y = msg->pose.pose.position.y;
  double car_z = msg->pose.pose.position.z;
  double yaw = msg->pose.pose.orientation.z;
  double v_x = msg->twist.twist.linear.x;
  double v_y = msg->twist.twist.linear.y;
  double v_z = msg->twist.twist.linear.z;

  double diff_radius = sqrt(pow(last_goal[0] - car_x,2) + pow(last_goal[1] - car_y,2) + car_z*car_z);
  double velocity = sqrt(pow(v_x, 2) + pow(v_y, 2) + pow(v_z, 2));

  if (diff_radius < 3 && !end) {
    ros::Duration(1).sleep();
    ROS_INFO("Duration: %f, Distance: %f, Energy: %f",duration, distance, energy);
    end = true;

    if (showgraph) {
      v_dat.data = v_array;
      a_dat.data = a_array;
      e_dat.data = e_array;
      d_dat.data = d_array;
      ma_dat.data = ma_array;
      gt_dat.data = goaltime;
      t_dat.data = t_array;

      while (pub_t.getNumSubscribers() == 0){
        if (ros::ok()) {
          ros::Duration(0.05).sleep();
        }
        else {
          return;
        }
      }
      pub_v.publish(v_dat);
      pub_a.publish(a_dat);
      pub_e.publish(e_dat);
      pub_d.publish(d_dat);
      pub_ma.publish(ma_dat);
      pub_gt.publish(gt_dat);
      ros::Duration(0.5).sleep();
      pub_t.publish(t_dat);
    }
    return;
  }
  else if (end) {
    return;
  }
  else if (start) {
    // double dv = velocity - v0;
    // v0 = velocity;
    double dv = sqrt(pow(v_x - v_x0, 2) + pow(v_y - v_y0, 2) + pow(v_z - v_z0, 2));
    v_x0 = v_x;
    v_y0 = v_y;
    v_z0 = v_z;

    double dd = sqrt(pow(car_x - x0, 2) + pow(car_y - y0, 2) + pow(car_z - z0, 2));
    distance += dd;
    x0 = car_x;
    y0 = car_y;
    z0 = car_z;

    // Calculate dt and save t0
    double t = ros::Time::now().toSec();
    double dt = t - t0;
    duration += dt;
    t0 = t;

    double acceleration = dv/dt;
    double f_d = drag * (velocity * velocity);
    double f_i = mass * acceleration;

    // calculate forces acting against the moving vehicle
    double f_total = f_r + f_d + f_i;
    double e = f_total*dd;
    energy += e;
    // ROS_INFO("duration is %f, dv is %f, dt is %f, dd is %f, energy is %f", duration, dv, dt_s, dd, energy);
    if (showgraph) {
      v_array.push_back(velocity);
      t_array.push_back(duration);
      a_array.push_back(acceleration);
      d_array.push_back(f_d);
      ma_array.push_back(f_i);
      e_array.push_back(e);
    }
  }
  else {
    if (velocity > 0) {
      t0 = ros::Time::now().toSec();
      start = true;
    }
  }
}
void Efficiency::getScore(const std_msgs::Int8::ConstPtr& num) {
  if (showgraph) {
    if (passed != num->data) {
      goaltime.push_back(duration);
      passed = num->data;
    }
  }
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "efficiency_calc_cpp");
  ros::NodeHandle n;
  Efficiency efficiency = Efficiency(&n);
  ros::spin();
  return 0;
}

File: AirSim/project/airsim_2dnav/scripts/efficiency.py
#!/usr/bin/env python
import rospy
import math
from std_msgs.msg import Float32, Float64, Int8
from nav_msgs.msg import Odometry
from matplotlib import pyplot as plt

# efficiency.py
# Author(s): Navaneeth Nair, Khai Hoe
# Documented by: Woon Siang Yi
# School: Monash University Malaysia
# Description: - Python script that calculates distance, duration and energy used by the car to travel from start to finish
#                by subscribing to the car's odometry data, it also calculates the cpu usage of car control and overall cpu usage by
#                subscribing to cpu_monitor
#              - Record the Velocity, Acceleration, Energy, Drag Force and Inertial Force of the car at a defined period
#                and plot several subgraphs consist of these data against time at the end

def odom(msg):
    global distance, energy, duration, start
    global x0, y0, z0, v0, t0, end
    global v_x0, v_y0, v_z0

    # Get car position, velocity and orientation
    car_x = msg.pose.pose.position.x
    car_y = msg.pose.pose.position.y
    car_z = msg.pose.pose.position.z
    v_x = msg.twist.twist.linear.x
    v_y = msg.twist.twist.linear.y
    v_z = msg.twist.twist.linear.z

    diff_radius = math.sqrt(math.pow(last_goal[0] - car_x,2) + math.pow(last_goal[1] - car_y,2) + car_z*car_z) # calculates the distance between the car and the final goal position using Pythagoras' Theorem
    velocity = math.sqrt(math.pow(v_x, 2) + math.pow(v_y, 2) + math.pow(v_z, 2)) # calculates the resultant velocity of the car

    if diff_radius < 3 and not end: # reached the end goal
        rospy.sleep(1)
        rospy.loginfo("Results: [Distance(m): %f, Duration(s): %f, Energy(J): %d, cpu_tot(avg): %f, cpu_tot(max): %f, cpu_cc(avg): %f,  cpu_cc(max): %f]" %(distance, duration, math.ceil(energy), cpu_avg, cpu_max, cc_avg, cc_max))
        end = 1

        if showgraph: # plot a graph if enabled
            #plt.figure(figsize=(8, 10))
            fig, ax = plt.subplots(figsize=(8,10))

            plt.subplot(5,1,1)
            plt.plot(t_array,v_array,'-r')
            plt.title("Velocity vs Time")

            plt.subplot(5,1,2)
            plt.plot(t_array,a_array,'-r')
            plt.title("Acceleration vs Time")

            plt.subplot(5,1,3)
            plt.plot(t_array,e_array,'-r')
            plt.title("Energy vs Time")

            plt.subplot(5,1,4)
            plt.plot(t_array,d_array,'-r')
            plt.title("Drag Force vs Time")

            plt.subplot(5,1,5)
            plt.plot(t_array,ma_array,'-r')
            plt.title("Inertial Force vs Time")

            plt.tight_layout()

            for p in fig.axes:
                pos = p.get_position()
                i = 0

                for t in goaltime:
                    p.axvline(t,ls ='--',color = 'g') # set line style and color
                    p.text(t,pos.x1,i) # add text to the axes
                    i += 1
            plt.show()
        return

    elif end:
        return

    elif start:
    	# Calculate velocity and save
    	# dv = velocity - v0
        # v0 = velocity
        dv = math.sqrt(math.pow(v_x - v_x0, 2) + math.pow(v_y - v_y0, 2) + math.pow(v_z - v_z0, 2))
        v_x0 = v_x
        v_y0 = v_y
        v_z0 = v_z

    	# Calculate distance travelled and save
    	dd = math.sqrt(math.pow(car_x - x0, 2) + math.pow(car_y - y0, 2) + math.pow(car_z - z0, 2))
    	distance += dd
    	x0 = car_x
    	y0 = car_y
    	z0 = car_z

    	# Calculate dt and save
    	t = rospy.get_time()
    	dt = t - t0
        duration += dt
    	t0 = t

        # Calculate acceleration
        acceleration = dv/dt

    	# Force Calculation
    	f_r = mass*g*friction # Road force
    	f_d = 0.5*rho*drag_coeff*area*math.pow(velocity,2) # Drag force
    	f_i = mass*acceleration # Inertial force

    	# Calculate Energy usage
    	f_tot = f_r + f_d + f_i # Total force
        e = f_tot*dd
    	energy += e

        if showgraph: # update arrays
            v_array.append(velocity)
            t_array.append(duration)
            a_array.append(acceleration)
            d_array.append(f_d)
            ma_array.append(f_i)
            e_array.append(e)
    else:
        t0 = rospy.get_time()
        if velocity > 0:
            start = True

# calculates car control cpu usage
def cpu_monitor_cc(usage):
	global cc, cc_cnt, cc_avg, cc_max, cc_cur
	cc_cur = usage.data

	if usage.data > cc_max:  # update maximum car control cpu usage
		cc_max = usage.data

    # calculates the average car control cpu usage
	cc_cnt = cc_cnt + 1
	cc = cc + usage.data
	cc_avg = cc/cc_cnt

# calculates overall cpu usage
def cpu_monitor_tot(usage):
	global cpu, cpu_cnt, cpu_avg, cpu_max

	if usage.data > cpu_max:  # update maximum overall cpu usage
		cpu_max = usage.data

    # calculates the average overall cpu usage
	cpu_cnt = cpu_cnt + 1
	cpu = cpu + usage.data
	cpu_avg = cpu/cpu_cnt

# records the duration to reach each goal in an array
def getscore(num):
    global passed, goaltime

    if showgraph:
        if (passed != num.data):
            goaltime.append(duration)
            passed = num.data

def listener():
    global goal_type
    rospy.init_node('efficiency_calc')
    rospy.Subscriber("/cpu_monitor/control_master/cpu", Float32, cpu_monitor_cc)
    rospy.Subscriber("/cpu_monitor/total_cpu", Float32, cpu_monitor_tot)
    rospy.Subscriber("/odom", Odometry, odom) # Get car position
    rospy.Subscriber("/current_score", Int8, getscore)
    rospy.spin()


if __name__ == '__main__':
    # Initialize variables
    goals = [[44.50, -65.75], [0, -256], [-84.5,-198.22], [-135, 1.50]] # Our common final goals
    last_goal = goals[3] # If using a different final goal, change here
    v0 = 0
    v_x0 = 0
    v_y0 = 0
    v_z0 = 0
    x0 = 0
    y0 = 0
    z0 = 0
    t0 = 0
    distance = 0
    duration = 0
    energy = 0
    end = False
    start = False

    # CPU
    cpu_cnt = 0
    cpu = 0
    cpu_avg = 0
    cpu_max = 0
    cc = 0
    cc_cnt = 0
    cc_avg = 0
    cc_max = 0

    # Energy constants
    mass = 2176.46 # Car mass [kg]
    g = 9.81 # Gravitational constant [m/s^2]
    friction = 0.01 # Rolling friction coefficient
    rho = 1.2 # Air density at NTP [kg/m^3]
    drag_coeff = 0.357 # Drag coefficient
    area = 3.389 # Car front chassis area [m^2]

    # Graph
    showgraph = True
    if (showgraph == True):
        v_array = []  # velocity array
        a_array = []  # acceleration array
        e_array = []  # energy array
        d_array = []  # drag force array
        ma_array = []  # inertial force array
        t_array = []  # time array
        goaltime = []
        passed = 0
    listener()
